xtask/src/run.rs

```diff
@@ -83,40 +102,383 @@ pub fn build(opts: BuildOptions) -> Result<Vec<(String, PathBuf)>> {
+                    match OS {
+                        "linux" => {
+                            let mut cmd = Command::new("locate");
+                            let output = cmd
+                                .arg("QEMU_EFI.fd")
+                                .output()
+                                .with_context(|| format!("failed to run {cmd:?}"))?;
+                            let Output { status, .. } = &output;
+                            if status.code() != Some(0) {
+                                bail!("{qemu:?} failed: {output:?}")
+                            }
+                            let Output { stdout, .. } = output;
+                            let bios = String::from_utf8(stdout)
+                                .with_context(|| format!("failed to parse output of {cmd:?}"))?;
+                            qemu.args(["-bios", bios.trim()]);
+                        }
+                        "macos" => {
+                            let mut cmd = Command::new("brew");
+                            let output = cmd
+                                .args(["list", "qemu", "-1", "-v"])
+                                .output()
+                                .with_context(|| format!("failed to run {cmd:?}"))?;
+                            let Output { status, .. } = &output;
+                            if status.code() != Some(0) {
+                                bail!("{qemu:?} failed: {output:?}")
+                            }
+                            let Output { stdout, .. } = output;
+                            let output = String::from_utf8(stdout)
+                                .with_context(|| format!("failed to parse output of {cmd:?}"))?;
+                            const NAME: &str = "edk2-aarch64-code.fd";
+                            let bios = output.lines().find(|line| line.contains(NAME)).ok_or_else(
+                                || anyhow!("failed to find {NAME} in output of {cmd:?}: {output}"),
+                            )?;
+                            qemu.args(["-bios", bios.trim()]);
+                        }
+                        os => bail!("unsupported OS: {os}"),
+                    };
+                }
+                let mut qemu_child = qemu
+                    .stdout(Stdio::piped())
+                    .spawn()
+                    .with_context(|| format!("failed to spawn {qemu:?}"))?;
+                let Child { stdout, .. } = &mut qemu_child;
+                let stdout = stdout.take().unwrap();
+                let stdout = BufReader::new(stdout);
+
+                let mut outcome = None;
+                for line in stdout.lines() {
+                    let line =
+                        line.with_context(|| format!("failed to read line from {qemu:?}"))?;
+                    println!("{}", line);
+                    if let Some(line) = line.strip_prefix("init: ") {
+                        let previous = match line {
```

Comment:
Some commenting around the expected outputs here might be helpful

Reply:
It's worth calling out that `init: success` and `init: failure` are hardcoded outputs from `init` -  a crate under out control. If they change, they must change here too.

`if let Some(previous) = previous {` seems to be an odd condition to reach.
Under which conditions could we set outcome more than once?

Reply:
Added callout. It's an unexpected condition we'd only reach if we had a bug.